SSA/Projects/CIRCT/Comb/Comb.lean

```diff
@@ -154,181 +140,84 @@ def ofList {Γ : Ctxt _} ty : (l : List ((ty : Comb.Ty) × Γ.Var ty)) → (h :
-          | "Comb.shl" =>
-            return ⟨_, .bv w₁,
-              (Expr.mk (op := .shl w₁) (ty_eq := rfl) (eff_le := by constructor)
-                (args := .cons v₁ <| .cons (h ▸ v₂) <| .nil) (regArgs := .nil) : Expr (Comb) Γ .pure (.bv w₁))⟩
-          | "Comb.shrs" =>
-            return ⟨_, .bv w₁,
-              (Expr.mk (op := .shrs w₁) (ty_eq := rfl) (eff_le := by constructor)
-                (args := .cons v₁ <| .cons (h ▸ v₂) <| .nil) (regArgs := .nil) : Expr (Comb) Γ .pure (.bv w₁))⟩
-          | "Comb.shru" =>
-            return ⟨_, .bv w₁,
-              (Expr.mk (op := .shru w₁) (ty_eq := rfl) (eff_le := by constructor)
-                (args := .cons v₁ <| .cons (h ▸ v₂) <| .nil) (regArgs := .nil) : Expr (Comb) Γ .pure (.bv w₁))⟩
-          | "Comb.sub" =>
-            return ⟨_, .bv w₁,
-              (Expr.mk (op := .sub w₁) (ty_eq := rfl) (eff_le := by constructor)
-                (args := .cons v₁ <| .cons (h ▸ v₂) <| .nil) (regArgs := .nil) : Expr (Comb) Γ .pure (.bv w₁))⟩
-          | _ => throw <| .generic s!"Unknown operation"
-        else
-          throw <| .generic s!"bitvector sizes don't match for '{repr opStx.args}' in {opStx.name}"
-      | _, _ => throw <| .generic s!"type mismatch in {opStx.name}"
-    | _ => throw <| .generic s!"expected two operands, found #'{opStx.args.length}' in '{repr opStx.args}'"
-  | op@"Comb.mux" =>
-    match opStx.args with
-    | v₁Stx::v₂Stx::v₃Stx::[] =>
-      let ⟨ty₁, v₁⟩ ← MLIR.AST.TypedSSAVal.mkVal Γ v₁Stx
-      let ⟨ty₂, v₂⟩ ← MLIR.AST.TypedSSAVal.mkVal Γ v₂Stx
-      let ⟨ty₃, v₃⟩ ← MLIR.AST.TypedSSAVal.mkVal Γ v₃Stx
-      match ty₁, ty₂, ty₃, op with
-      | .bv w₁, .bv w₂, .bool, "Comb.mux" =>
-        /- mux currently only works if w₁ = w₂, since we need to fix the output type of the operation
-          it should work even if w₁ ≠ w₂ but i need to think about how to implement that in an elegant way -/
-        return ⟨_, .typeSum w₁ w₂,
-          (Expr.mk (op := .mux w₁ w₂) (ty_eq := rfl) (eff_le := by constructor)
-            (args := .cons v₁ <| .cons v₂ <| .cons v₃ <| .nil) (regArgs := .nil) : Expr (Comb) Γ .pure (.typeSum w₁ w₂))⟩
-      | _, _, _, _ => throw <| .generic s!"type mismatch"
-    | _ => throw <| .generic s!"expected three operands, found #'{opStx.args.length}' in '{repr opStx.args}'"
-  |  _ =>
+  let args ← opStx.parseArgs Γ
+  -- 1-ary ops
+  let unW := do
+    let args ← args.assumeArity 1
+    return getVarWidth args[0]
+  -- 2-ary ops
+  let binW : AST.ReaderM (Comb) (Nat) := do
+    let args ← args.assumeArity 2
+    return getVarWidth args[0]
+  -- 3-ary ops
+  let terW : AST.ReaderM (Comb) (Nat) := do
+    let args ← args.assumeArity 3
+    return getVarWidth args[0]
+  -- n-ary ops
+  let args' ← opStx.args.mapM (MLIR.AST.TypedSSAVal.mkVal Γ) -- will need to find a better way to do this
+  if h : args'.length = 0 then
+    throw <| .generic s!" empty list of argument provided for the variadic op {repr opStx.args}"
```

Comment:
NIT: Note that this check is performed eagerly, so IIUC, even for non-variadic ops if you pass an empty list then it will throw this error

Reply:
there is always at least an argument